src/checkpointing.jl

```diff
@@ -0,0 +1,29 @@
+"""
+    checkpoint(f, args::Tuple)
+
+Treat `f` as a primitive, whose gradient is computed by running and forwards- and reverse-
+pass of ∇ inside the outer pass. This can have the effect of significantly lowering the
+peak memory usage of a gradient computation at the expense of additional computation.
+
+HEALTH WARNING: If you close over differentiable objects in `f`, you will get incorrect
```

Comment:
> HEALTH WARNING

:joy:

Reply:
We could detect if something is a closure and print a warning.
Harder to detect if is is a closure over something differentiable.
But being a closure at all can be detected by

`fieldcount(typeof(f)) > 0 && f isa Function`

Reply:
Ooo it hadn't dawned on me that this is a thing that would work. Will implement. I don't think that we need to check that `f isa Function` though, since callables needn't subtype `Function`.